docs: add glossary, `add` explainer, and content routing FAQ (#233)
filecoin-project/filecoin-pin: documentation/glossary.md

Filecoin Pin brings multiple technologies together (e.g., traditional Filecoin blockchain, traditional IPFS, new Filecoin initiatives like Filecoin Onchain Cloud).  As a result, terminology from all these areas is used for describing Filecoin Pin.  This glossary serves as a primer of the key terminology.  Rather than seeking to be the comprehensive source of truth, it seeks to point to where to find authoritative and more in depth information.  Many additional IPFS-related terms can be found in https://docs.ipfs.tech/concepts/glossary.  

## Calibration Network

The “test network” of Filecoin’s Mainnet, where developers can have more realistic network conditions without using truly valuable tokens.  See https://docs.filecoin.io/networks/calibration

## CAR

> A CAR is a container and transport to hold your "IPFS data" (i.e., IPLD blocks).  It just happens to take file form sometimes.  See https://docs.ipfs.tech/concepts/glossary/#car for more info.

Reviewer: haha, love it, this sounds like my words

```suggestion
A CAR is a container format and network transport to hold your "IPFS data" (i.e., IPLD blocks).  It just happens to take file form sometimes so is often called a "CAR file".  See https://docs.ipfs.tech/concepts/glossary/#car for more info.
```

This is often abbreviated as “FOC”, which yes, does phonetically resonate with more colorful language 😉.

## Filecoin Pin

https://github.com/FilOzone/filecoin-pin

Serves as an IPFS-oriented set of tools for interfacing with [Filecoin Onchain Cloud](#filecoin-onchain-cloud) built on top of [Synapse](#synapse).

## `filecoin-pin`

`filecoin-pin` is a CLI tool affordance for [Filecoin Pin](#filecoin-pin).

## filecoin-pin-website

https://github.com/filecoin-project/filecoin-pin-website

Example of [Filecoin Pin](#filecoin-pin) in action within a web-browser.  Its purposes are:

1. Demonstrate that Filecoin Pin is usable.  Drag and drop and you're good to go!
2. Serve as a starter or inspiration for dApp builders wanting to use [Filecoin Onchain Cloud](#filecoin-onchain-cloud).

filecoin-pin-website is also hosted at [pin.filecoin.cloud](http://pin.filecoin.cloud), with hardcoded wallet and [session key](#session-key) on the [Calibration](#calibration-network) network.  In future, [integration with tools like Metamask will be supported](https://github.com/filecoin-project/filecoin-pin-website/issues/77).  

## Filecoin Warm Storage Service

> This is the entry point smart contract for using the warm storage functionality offered in [Filecoin Onchain Cloud](#filecoin-onchain-cloud).  It acts as a validator for payment settlements, ensuring the warm storage service is actually delivered before payments are released to the [Storage Provider](#storage-provider).

Reviewer: ```suggestion
This is the primary smart contract used when interacting with the warm storage functionality offered in [Filecoin Onchain Cloud](#filecoin-onchain-cloud).  It acts as both a "service" contract and a "validator" contract for payment management and settlements, ensuring the warm storage service is actually delivered before payments are released to the [Storage Provider](#storage-provider).
```

## IPFS Root CID

> The CID for the root of a merkle DAG that is usually encoding a file or directory as UnixFS.  Since each `filecoin-pin add` creates a [CAR](#car), regardless if passed a file or directory, there is a single root corresponding to root of the Merkle DAG made out of encoding the file or directory as UnixFS.

Reviewer: ```suggestion
The CID for the root of a merkle DAG that is usually encoding a file or directory as UnixFS.  Since each `filecoin-pin add` creates a [CAR](#car), regardless if passed a file or directory, there is a single root corresponding to root of the Merkle DAG made out of encoding the file or directory as UnixFS. Typically this will be presented in base32, beginning with `bafy` and be 59 characters long.
```

## `/ipfs` Retrieval

> This is one of two retrieval endpoints that [Storage Providers](#storage-provider) expose.  This endpoint conforms with the [IPFS Trustless Gateway Specification](https://specs.ipfs.tech/http-gateways/trustless-gateway/).  All CIDs that are indexed by the SP should be retrievable via this endpoint.  This is the endpoint that is announced through the provider records stored by [IPNI](#ipni) Indexers.

Reviewer: ```suggestion
This is one of two retrieval endpoints that [Storage Providers](#storage-provider) expose.  This endpoint conforms with the [IPFS Trustless Gateway Specification](https://specs.ipfs.tech/http-gateways/trustless-gateway/).  All CIDs that are indexed by the SP should be retrievable via this endpoint.  This is the endpoint that is announced through the provider records stored by [IPNI](#ipni) Indexers. As a "trustless" protocol, retrieval of IPFS data using this mechanism provides assurance that data has not been tampered with and that what is being retrieved is _exaclty_ what was requested. This is in contrast to a "trusted" gateway where IPFS data is reassembled into a form appropriate for rendering. Developers and users are encouraged to perform this reassembly step as close as possible to the user, using existing IPFS technologies such as [Kubo](https://github.com/ipfs/kubo) and [Helia](https://github.com/ipfs/helia). For example, Helia's [`verified-fetch` package](https://www.npmjs.com/package/@helia/verified-fetch) is able to perform this within a browser context, this is the technology that powers https://inbrowser.link/.
```

## Piece

A Piece is an individual unit of data identified by [Piece CID](#piece-cid). Multiple Pieces can be added to a [Data Set](#data-set) for storage.

With [Filecoin Pin](#filecoin-pin), the Piece is the [CAR](#car) file itself; an array of bytes representing the serialized content. Each `filecoin-pin add` operation creates exactly one Piece by converting the input file or directory to a CAR file, which then becomes the Piece that is uploaded and stored.

## Piece CID

A CID for a [Piece](#piece) using the [CommP](#commp) hash function.  This is a common CID type used within Filecoin.  This value is different than the [IPFS Root CID](#ipfs-root-cid).  This is the CommP hash for the full [CAR](#car) itself while serially processing its bytes without any regard to the CAR's underlying DAG structure.

## `/piece` Retrieval

This is a Filecoin-defined retrieval specification outlined in https://github.com/filecoin-project/FIPs/blob/master/FRCs/frc-0066.md.  It is for retrieving pieces by [Piece CID](#piece-cid), optionally taking a range.

It takes the form of https://sp.domain/piece/$pieceCid.

## Proof of Data Possession

https://github.com/FilOzone/pdp

> The cryptographic protocol that verifies [storage providers](#storage-provider) are actually storing the data they claim to store. Providers must periodically prove they possess the data.  

Reviewer: ```suggestion
The cryptographic protocol that verifies [storage providers](#storage-provider) are actually storing the data they claim to store. Providers must periodically prove they possess the data.  This is distinct from the existing Filecoin proof system, "PoRep" or "Proof of Replication".
```

This is usually abbreviated as “PDP”.

## RPC Provider

HTTP endpoint/infrastructure for reading or writing blockchain state.  These RPC providers run native blockchain clients and likely are storing blockchain state in an optimized format for faster reads.  See https://docs.filecoin.io/networks/mainnet/rpcs for for more information about Filecoin RPC providers. 

  

## Session Key

> An onchain registry of [Storage Providers](#storage-provider) who are participating in [Filecoin Onchain Cloud](#filecoin-onchain-cloud).  They can be viewed at https://filecoin.cloud/providers.  By default, only "Approved Providers" are used by [Filecoin Pin](#filecoin-pin) because they have been vetted to support IPFS Mainnet retrievals.

Reviewer: ```suggestion

## Service Provider Registry

An onchain registry of [Storage Providers](#storage-provider) who are participating in [Filecoin Onchain Cloud](#filecoin-onchain-cloud).  They can be viewed at https://filecoin.cloud/providers.  By default, only "Approved Providers" are used by [Filecoin Pin](#filecoin-pin) because they have been vetted to support IPFS Mainnet retrievals.
```
